builder/qemu/config.go

```diff
@@ -65,6 +65,96 @@ type QemuImgArgs struct {
 	Resize  []string `mapstructure:"resize" required:"false"`
 }
 
+// QemuSMPConfig sets the smp configuration option for the Qemu command-line
+//
+// The smp option sets the number of vCPUs to expose to the VM, the final
+// number of available vCPUs is sockets*cores*threads.
+type QemuSMPConfig struct {
+	// The number of virtual cpus to use when building the VM.
+	//
+	// If undefined, the value will either be `1`, or the product of
+	// `sockets * cpus * threads`
+	//
+	// If both any topology element, or this is set, the final value will be the
+	// largest of the two.
+	//
+	// If the cpu count is the only thing specified, qemu's default behaviour
+	// regarding topology will be applied.
+	// The behaviour depends on the version of qemu; before version 6.2, sockets
+	// were preferred to cores, from version 6.2 onwards, cores are preferred.
+	CpuCount int `mapstructure:"cpus" required:"false"`
+	// The number of sockets to use when building the VM.
+	//  The default is `1` socket.
+	//  The socket count must not be higher than the CPU count.
+	SocketCount int `mapstructure:"sockets" required:"false"`
+	// The number of cores per CPU to use when building the VM.
+	//  The default is `1` core per CPU.
+	CoreCount int `mapstructure:"cores" required:"false"`
+	// The number of threads per core to use when building the VM.
+	//  The default is `1` thread per core.
+	ThreadCount int `mapstructure:"threads" required:"false"`
+}
+
+// getDefaultCmdLine returns the smp command-line argument if defined.
+func (c QemuSMPConfig) getDefaultCmdLine() string {
+	totalVCpus := c.getMaxCPUs()
+	if totalVCpus == 0 && c.CpuCount == 0 {
+		return "1"
+	}
+
+	cpuCount := c.CpuCount
+
+	if cpuCount == 0 {
+		log.Printf("CPU count at default value, setting to topology maximum: %d", totalVCpus)
+		cpuCount = totalVCpus
+	}
+
+	if totalVCpus > cpuCount {
+		log.Print("CPU count lower than what's described in topology." +
+			"This will negatively impact performance.")
+	}
+
+	if cpuCount > totalVCpus && totalVCpus != 0 {
+		log.Printf("CPU count greater than what topology allows, setting to %d", totalVCpus)
+		cpuCount = totalVCpus
+	}
+
+	smpStr := fmt.Sprintf("%d", cpuCount)
+	if c.SocketCount > 0 {
+		smpStr = fmt.Sprintf("%s,sockets=%d", smpStr, c.SocketCount)
+	}
+	if c.CoreCount > 0 {
+		smpStr = fmt.Sprintf("%s,cores=%d", smpStr, c.CoreCount)
+	}
+	if c.ThreadCount > 0 {
+		smpStr = fmt.Sprintf("%s,threads=%d", smpStr, c.ThreadCount)
+	}
+
+	return smpStr
+}
+
+func (c QemuSMPConfig) getMaxCPUs() int {
```

Comment:
Even though this is a unexported method I recommend adding comment to explain how getMaxCPUs works. More specifically that it calculates the max cpu based off the QemuSMPConfig values. At first I thought that it was making a call to the underlying system to determine max allowed cpu.